### PR TITLE
Thumbnail Setter Addon

### LINKED ISSUE #67: [New addon] Animated/custom thumbnail setter by World_Languages
**Describe the new addon**

My animated thumbnail bookmarklet as an addon.

**Link to the new addon**

https://github.com/WorldLanguages/animatedThumbnailsBookmarklet/blob/master/code.js

Should probably be off by default, and add a button so it can be used if enabled.

### PR DESCRIPTION
**Pull request by TheColaber**
**Resolves**

Resolves #67

**Reason for changes**

It is very useful to set your project's thumbnail to an image that may be a gif.

**Tests**

Works on chrome. I know that there will be a lot of changes that would need to be made to the userscript.js file so if anyone can help, that would be great!

### CODE REVIEW

File: addons/animated-thumb/userscript.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      var script = document.createElement('script');
+      script.src = 'https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js';
+      script.type = 'text/javascript';
+      script.onload = animThumbnailMain;
+      document.getElementsByTagName('head')[0].appendChild(script);
+
+      function animThumbnailMain() {
+        let snackBarCSS = function() {
+          var css = document.createElement("style");
+          css.innerHTML = '#snackbar { visibility: hidden; min-width: 250px; margin-left: -125px; background-color: black; color: #fff; text-align: center; border-radius: 2px; padding: 16px; position: fixed; z-index: 1; left: 50%; top: 50px; } #snackbar.show { visibility: visible; } ';
```
Comment (TheColaber) [inline addons/animated-thumb/userscript.js:30]: What can replace this .innerHTML with?

Comment (TheColaber) [reply addons/animated-thumb/userscript.js:30]: or should I create a userstyle?

Comment (JeffaloBob) [reply addons/animated-thumb/userscript.js:30]: a userstyle

Comment (TheColaber) [reply addons/animated-thumb/userscript.js:30]: ok.

**Comment by TheColaber** [comment]
Idk what to replace with the other .innerHTMLs... Does anyone know?

Also, can someone review it for the other stuff besides the .innerHTMLs?

File: addons/animated-thumb/addon.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "Thumbnails Setter",
+  "description": "This addon will allow you to set the thubnail of your project to any photo (including gif images)",
```
Comment (mxmou) [inline addons/animated-thumb/addon.json:3]: Maybe "any image", because it doesn't need to be a photo?

**Comment by mxmou** [comment]
> Idk what to replace with the other .innerHTMLs... Does anyone know?

Setting `innerHTML` to a constant can always be replaced by multiple calls to `createElement`, `createTextNode` and `appendChild`, but I'm not sure if that's necessary here. @WorldLanguages should tell that.

**Comment by Explosion-Scratch** [comment]
That's pretty much the same thing but harder? Why do the extension stores
care?

File: addons/animated-thumb/addon.json (hunk 2 above)
Comment (TheColaber) [reply addons/animated-thumb/addon.json:3]: I'll do that...

**Comment by TheColaber** [comment]
> > Idk what to replace with the other .innerHTMLs... Does anyone know?
> 
> Setting `innerHTML` to a constant can always be replaced by multiple calls to `createElement`, `createTextNode` and `appendChild`, but I'm not sure if that's necessary here. @WorldLanguages should tell that.

Well I know I can't use `createTextNode` due to there being `<br>`s and `<a>`s
I can't use `createElement` because there is text and other elements in one. However, I could create a bunch of `span` elements and those `<br>`s and `<a>`s but that would be a lot of work. And `appendChild` appends a Node. Text can't be a node.

@WorldLanguages Thoughts?

**Comment by mxmou** [comment]
> And `appendChild` appends a Node. Text can't be a node.

It can be - that's what `createTextNode` is for.

**Comment by TheColaber** [comment]
> > And `appendChild` appends a Node. Text can't be a node.
> 
> It can be - that's what `createTextNode` is for.

Oh yeah...

**Comment by GrahamSH-LLK** [comment]
A lot of the elements that you are using include things like _3wjWH. These change with every build of scratch, and will break your addon. Replace them with [attribute^="beginning"], for example ".menu-bar_main-menu_3wjWH" to [class^="menu-bar_main-menu"] I'll fix some of these myself.

**Comment by TheColaber** [comment]
> A lot of the elements that you are using include things like _3wjWH. These change with every build of scratch, and will break your addon. Replace them with [attribute^="beginning"], for example ".menu-bar_main-menu_3wjWH" to "[class^="menu-bar_main-menu"]" I'll fix some of these myself.

Hmm, I don't quite understand... could you do fix this for me?

**Comment by TheColaber** [comment]
> A lot of the elements that you are using include things like _3wjWH. These change with every build of scratch, and will break your addon. Replace them with [attribute^="beginning"], for example ".menu-bar_main-menu_3wjWH" to "[class^="menu-bar_main-menu"]" I'll fix some of these myself.

Like do you mean replace this line:
```javascript
let nav = await addon.tab.waitForElement(".menu-bar_main-menu_3wjWH", {markAsSeen: true});
```
with
```javascript
let nav = await addon.tab.waitForElement("[class^="menu-bar_main-menu"]", {markAsSeen: true});
```

**Comment by TheColaber** [comment]
> > A lot of the elements that you are using include things like _3wjWH. These change with every build of scratch, and will break your addon. Replace them with [attribute^="beginning"], for example ".menu-bar_main-menu_3wjWH" to "[class^="menu-bar_main-menu"]" I'll fix some of these myself.
> 
> Like do you mean replace this line:
> 
> ```js
> let nav = await addon.tab.waitForElement(".menu-bar_main-menu_3wjWH", {markAsSeen: true});
> ```
> 
> with
> 
> ```js
> let nav = await addon.tab.waitForElement("[class^='menu-bar_main-menu']", {markAsSeen: true});
> ```

Ohhhhh I see now, yeah, I will do that now.

File: addons/animated-thumb/addon.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "Thumbnails Setter",
```
Comment (WorldLanguages) [inline addons/animated-thumb/addon.json:2]: Naming conventions

File: addons/animated-thumb/addon.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  "description": "This addon will allow you to set the thubnail of your project to any image (including gif images)",
```
Comment (WorldLanguages) [inline addons/animated-thumb/addon.json:3]: 1. Typo "thubnail"
2. "This addon will allow you to set the thumbnail of your project" how? You should mention where the button is placed.
3. Repetition of word "image"
4. Missing dot at the end

File: addons/animated-thumb/addon.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  "credits": [
+    {
+      "name": "World_Languages",
+      "link": "https://github.com/WorldLanguages/animatedThumbnailsBookmarklet/blob/master/code.js"
+    },
+    {
+      "name": "TheColaber",
+      "link": "https://scratch.mit.edu"
```
Comment (WorldLanguages) [inline addons/animated-thumb/addon.json:11]: Wrong link?

Comment (TheColaber) [reply addons/animated-thumb/addon.json:11]: Oh, wow... I need to sleep...

**Comment by WorldLanguages** [comment]
Can you make the button not work unless the project is owned by the user? You can check if the title is editable.

**Comment by WorldLanguages** [comment]
Also, I think you should not port my code into the addon itself, but make the addon load the code from the URL directly, exactly like the bookmarklet does. If my code breaks, then I wouldn't need to change it twice.

**Comment by TheColaber** [comment]
> Also, I think you should not port my code into the addon itself, but make the addon load the code from the URL directly, exactly like the bookmarklet does. If my code breaks, then I wouldn't need to change it twice.

Ok, but I changed some of your code so I could get rid of the `.innerHTML`s, so will you need to update your code or would the webstore be fine with that? I did do some other changes to so it would be easier if you update your code.

**Comment by WorldLanguages** [comment]
The `innerHTML` warnings only appear if they are within the extension source code so that's not an issue.
Please load the script with `addon.tab.loadScript(url)`

**Comment by TheColaber** [comment]
> The `innerHTML` warnings only appear if they are within the extension source code so that's not an issue.
> Please load the script with `addon.tab.loadScript(url)`

Sure!

**Comment by TheColaber** [comment]
> addon.tab.loadScript(url)

Hmm, I'm getting this error
```javascript
Refused to execute script from 'https://raw.githubusercontent.com/WorldLanguages/animatedThumbnailsBookmarklet/master/code.js' because its MIME type ('text/plain') is not executable, and strict MIME type checking is enabled.
```
I'm probably not doing this right...

**Comment by WorldLanguages** [comment]
@TheColaber I told you to check how the bookmarklet did it. The bookmarklet imports `https://worldlanguages.github.io/animatedThumbnailsBookmarklet/code.js`

**Comment by TheColaber** [comment]
> @TheColaber I told you to check how the bookmarklet did it. The bookmarklet imports `https://worldlanguages.github.io/animatedThumbnailsBookmarklet/code.js`

Oh, I didn't see that from before...

**Comment by GrahamSH-LLK** [comment]
This is related to https://github.com/LLK/scratch-www/issues/4092.

**Pull request merged**